### ISSUE TITLE
Timer processing in RC310

### ISSUE THREAD

**Issue by proddy** [emsesp/EMS-ESP32#1594]
### Discussed in https://github.com/emsesp/EMS-ESP32/discussions/1593

<div type='discussions-op-text'>

<sup>Originally posted by **Nxtway** January 30, 2024</sup>
ems-esp supports timer processing of the RC35, [#979](https://github.com/emsesp/EMS-ESP32/discussions/979)
Could these functionality also implemented for the RC310, please ? 

The type-id's in RC310 are:
circulation timer: thermostat(0x10) type(0x0309) | ON - FF | OFF - 00
warm water timer: thermostat(0x10) type(0x02FF) | COMFORT - 02 | ECO - 01 | OFF - ??, maybe 00
heating timer 1: thermostat(0x10) type(0x0683)
heating timer 2 thermostat(0x10)  type(0x068D)

Log files attached:
[WW_timer.txt](https://github.com/emsesp/EMS-ESP32/files/14103769/WW_timer.txt)
[circulation_timer.txt](https://github.com/emsesp/EMS-ESP32/files/14103770/circulation_timer.txt)
[heating_timer1.txt](https://github.com/emsesp/EMS-ESP32/files/14103771/heating_timer1.txt)
[heating_timer2.txt](https://github.com/emsesp/EMS-ESP32/files/14103772/heating_timer2.txt)

An additional improvement might be the holiday timer, but I didn't find the type-id up to now.</div>

**Comment by MichaelDvP** [comment]
Oh, the heating timer for RC300 are 0x1C3-ff, and 0x349-ff. with mode instead of temperature setting. We have to split RC300 - RC310 - BC400. But product-id 158 includes RC300 and RC310, maybe second identifier (unknown).

But as mentioned in the lot of discussions before, i don't know how we get that much data in ram and mqtt.
Each program has 42 entries with Day-of-week, time and mode/temperature for progA, progB, DHW, circulation:
504 entities to display, holiday modes are in similar range.
For RC35 we use only a single switchpoint to show/edit with a command to change the index, but this way was always not accepted by the users.

**Comment by Nxtway** [comment]
yes, I can understand that argument.
Does it still make sense to raise the RC310 to the same level as the RC35 or leave it as it is?
![image](https://github.com/emsesp/EMS-ESP32/assets/82314801/24263c61-a47e-4f34-82f3-7194ceb76723)

**Comment by Nxtway** [comment]
I agree. It's would be nice to display the timer entries in ems-esp but changing them is not the use-case. 
The timers can enabled or disabled with the wwmode, wwcircmode, etc. . 
You can postpone this feature request

**Comment by Nxtway** [comment]
And here are the addresses of the vacation timers in the RC310:
![image](https://github.com/emsesp/EMS-ESP32/assets/82314801/9da38950-a139-4d77-8217-699445b93c9c)

vacation1: 
000+04:41:20.516 T 386: [emsesp] Me(0x0B) -R-> thermostat(0x10), ?(0x0269), length: 0x19
000+04:41:20.580 N 388: [emsesp] thermostat(0x10) -W-> Me(0x0B), ?(0x0269), data: 18 03 08 18 03 0E 02 11 02 FF 00 00 00 00 00 00 00 FF 00
data 0..2, start: 2024.03.08 ---> 18 03 08
data 3..5, end  : 2014.03.14 ---> 18 03 0E
data 6, heating mode: 0x01=as Saturday; 0x02=const temp; 0x03= off; 0x04=reduce
data 7, heating temp: 0x11 = 17°C
data 8, hot water: 0x01=as Saturday; 0x02=off: 0x03=off with disinfect
data 9, used for heating: FF=YES | 00=NO
data 17,used for hotwater: FF=YES | 00=NO

with a few lines of JS code in node-red it looks like:
![image](https://github.com/emsesp/EMS-ESP32/assets/82314801/f19efd03-ef1c-46f1-900d-a3799e56e9d0)

**Comment by proddy** [comment]
what shall we do here @MichaelDvP ?

**Comment by proddy** [comment]
I don't have these features on my thermostat so not sure how to proceed. I'll remove this issue from the backlog.

**Comment by tp1de** [comment]
I am using a Buderus gas boiler with RC310 thermostat and 2 heating circuits I can decode 4 switchTimes per hc.
Each hc has 2 switchPrograms A and B and 2 variants with level and temps. Dhw has another 2 switchPrograms.
So in total I have 10 with 2 hc's. (Would be up to 18 with 4 hc's)

For dhw there are 2 switchPrograms:
01FF for warm water buffer load and 0209 for circulation pump - switch between eco/hot and on/off

The heating circuits can be set to use levels (eco/comfort) or absolute temps per switchtime.
These are different telegram types the to be used. (see below all types are raw)

The switch is done by the entity hcx/switchprogmode for each heating circuit between levels and absolute (temps).
These entities have been build-in just recently.

switchProgram A: hc1 telegrams level/temps are: 01C3 / 0583 and hc2: 01C4 / 0584
switchProgram B: hc1 telegrams level/temps are: 0349 / 058D and hc2: 034A / 058E

Within #1934 long custom entities in raw are now realized and could be used to maintain the long 84 bytes telegrams.
Open questions is where to deconstruct these telegrams and which home automation systems should be supported.


Since quite some time I implemented this in my ioBroker adapter in read/write mode creating the same JSON-objects as the original km200 gateway does. But I needed to implement the right timing for getting the long telegrams and writing them back.
It is much better to do this in the EMS-ESP gateway.

**Comment by tp1de** [comment]
Each switchProgram has 6 switchTimes per day, 42 switchTimes per Week combined with 42 values. 
My 10 switchPrograms would create up to 420 mqtt entities  if 1:1 implemented and enabled. (2 hc's and dhw with circ. pump)

**Comment by MichaelDvP** [comment]
@tp1de  Please use the ems-esp notation for telegram-types. e.g. 0x2FF for dhw1. Thanks.

Can you confirm the value order:
first byte state, second byte time in 15min steps. --- time 0xFF stands for `not_set`?  (i'm not sure)
first 6 pairs for monday, next 6 for tuesday, etc. (or other day order?) -- are the 6 pairs  in row or can only first and last pair be set?
state is for: 
dhw1, type 0x2FF, 0x01 for `low`, 0x02 for `high`  --- other states, eg. 0 for off?
circulation dhw1, type 0x309, 0xFF for `on`, 0x00 for `off`
switchtime1 hc1 type 0x2C3, 0x03 for `eco`, --- **?** for `comfort` -- are there other values?
switchtime2 hc1 type 0x683, temperature in 0,5°C steps  -- is there min/max? 

switchtime 1/2 is selected by program, 0x2B9, offset 11
mode/temperature telegrams (2C3/683) is selected by switchprogmode, offset 19
 
We can read the whole telegram in ram and generate a json on API command. 
For RC35 i have done this a while ago to test. The byte order is different, RC35 have 42 switchtimes and each have also coded the day of week, there is no order in the settings, eg. offset 0/1 can be Fr evening, offset 6/7 mon morning, and 2-5 not set.
Here is the output for `ems-esp.local/api/thermostat/hc2/switchtime`
[RC35_timer.txt](https://github.com/user-attachments/files/16589637/RC35_timer.txt)

for RC300 i think the json should look 
```
"value":{
  "mo":{
    "01:{
      "time":"07:00","mode":"comfort"
    },
    "02": {
       ...
    }
  },
  "tu":{
  }
}
```

For commands we can use the same json structure ans json object in value field. 
Maybe only a single switchpoint per command, or more, to be discussed if additive or overwriting. to not set values.

But i have no idea 
- how we can publish it to mqtt
- what HA needs to interprete this
- how to display in webUI 

We need a complete concept before starting to code.

**Comment by tp1de** [comment]
> Please use the ems-esp notation for telegram-types. e.g. 0x2FF for dhw1. Thanks.

EMS+ and EMS2.0 switchPrograms (For me RC310 and Moduline 2050)

1st step is to select the right telegram type defined by reading the new entities switchprogmodes.

hc1 - Mode A   level 02C3  and absolute temps 0683 (A == program(1))
hc1 - Mode B   level 0449  and absolute temps 068D (B == program(2))

hc2 - Mode A   level 02C4  and absolute temps 0684 (A == program(1))
hc2 - Mode B   level 044A  and absolute temps 068E (B == program(2))

dhw1 - Mode A   02FF
dhw1 - cp (circulation pump) 0309

Each telegram has a length of 84 bytes:
per day there is a maximum of 6 switch points. Each switchpoint is represented by 2 bytes. Therefore each day is 12 bytes:
Mo offset 0  0x00
Tu offset 12 0x0C
We offset 24 0x18
Th offset 36 0x24
Fr offset 48 0x30
Sa offset 60 0x3C
Su offset 72 0x48

Telegram structure of one day looks like this with 2 active switchpoints:
P1 P2 P3 P4 P5 P6 ...........................P12
14 01 58 03 FF 03 FF 03 FF 03 FF 03

P1: time of 1st sp: unit 15 minutes --> 0x14 = 20 *15 = 300 minutes = 05:00 hours
FF: no active switchpoint

P2 are either levels or absolute temps depending on type.
Levels:  "FF": "on" / "03": "comfort" / "02": "high" / "01": "eco/low" /  "00": "off"
temps:  Number("0x" + hex) / 2

**Comment by tp1de** [comment]
km200 JSON uses time in minutes:
[
  {
    "dayOfWeek": "Mo",
    "setpoint": "high",
    "time": 330
  },
  {
    "dayOfWeek": "Mo",
    "setpoint": "low",
    "time": 600
  },
  {
    "dayOfWeek": "Mo",
    "setpoint": "high",
    "time": 1050
  },
  {
    "dayOfWeek": "Mo",
    "setpoint": "low",
    "time": 1260
  },
  {
    "dayOfWeek": "Tu",
    "setpoint": "high",
    "time": 330
  },
  ..........
]

**Comment by tp1de** [comment]
Within ioBroker I write times:
[
  {
    "dayOfWeek": "Mo",
    "setpoint": "high",
    "time": "5:30"
  },
  {
    "dayOfWeek": "Mo",
    "setpoint": "eco/low",
    "time": "10:00"
  },
  {
    "dayOfWeek": "Mo",
    "setpoint": "high",
    "time": "17:30"
  },

**Comment by tp1de** [comment]
I think we do not need the index 01, 02 etc ... in JSON

**Comment by tp1de** [comment]
The API version with JSON will deliver the similar data object compared to the original Bosch API. 
Using MQTT for HA I can not identify the right object yet.

At the end JSON structure is not nice to manipulate - neither in ioBroker nor in HA.
I am reading the km200 data with node-red as well and create the entities by using mqtt discovery.
This works nice an it would be possible to create single entities for all switchpoints. while reading the JSON.
Today the switchPrograms are just read-only sensors with JSON-content (text).

**Comment by MichaelDvP** [comment]
> Telegram structure of one day looks like this with 2 active switchpoints:
P1 P2 P3 P4 P5 P6 ...........................P12
14 01 58 03 FF 03 FF 03 FF 03 FF 03

Ah, ok, first time, second mode. (RC35 have reverse order).
My question is: is this ordered? Or is it possible to have two setpoints this way (second and last set, others unset, last is before second):
`FF 03 58 03 FF 03 FF 03 FF 03 14 01`
> I think we do not need the index 01, 02 etc ... in JSON

If it is not ordered and we want to skip all unset points, we need the number to edit/delete a position. Or we always have to publish/write all unset values.
> Levels: "FF": "on" / "03": "comfort" / "02": "high" / "01": "eco/low" / "00": "off"

Does this affect all? Is it really possible to set a cirulation pump to "comfort", or heatingscircuit to "on"? 
Or (what i think is more logical) does it depend on telegram: hc: comfort/eco, dhw: high/low, circulation on/off?
> temps: Number("0x" + hex) / 2

Again: is there a limit? Is it possible to set temp to 3 (1.5°C) or 0xFF (127,5°C)? Or can we use the number to show mode for FF, 3, 2, 1, 0 and temp for all other values.

**Comment by tp1de** [comment]
> Ah, ok, first time, second mode. (RC35 have reverse order).
> My question is: is this ordered? Or is it possible to have two setpoints this way (second and last set, others unset, last is before second):

I have never seen non-ordered telegrams. An index is not necessary and is not used within original Bosch API
All unset values use "0xFF"  for time and in my case "03" for value. (hc's with level). Temp values can be between 5 (off) and 30.

> If it is not ordered and we want to skip all unset points, we need the number to edit/delete a position. Or we always have to publish/write all unset values.

I do always publish a full day if changed (12 bytes including unset points) - So max 7 telegrams

> Levels: "FF": "on" / "03": "comfort" / "02": "high" / "01": "eco/low" / "00": "off
Does this affect all? Is it really possible to set a cirulation pump to "comfort", or heatingscircuit to "on"?
Or (what i think is more logical) does it depend on telegram: hc: comfort/eco, dhw: high/low, circulation on/off?

Decoding for reading telegram is as defined. But possible enum values for writing are:
hc level: "eco/low", and "comfort" (for my gas boiler - we might need to check heatpumps maybe "off" as well")
dhw (buffer):  "eco/low" and "high" - maybe "off" for other systems ??
dhw cp:  "off" and "on"

> Again: is there a limit? Is it possible to set temp to 3 (1.5°C) or 0xFF (127,5°C)? Or can we use the number to show mode for FF, 3, 2, 1, 0 and temp for all other values.

Minimum temp is 5°C (off) and maximum is 30°C - the same for manual hc values.

**Comment by MichaelDvP** [comment]
> The switchtimes do not deliver any values / results yet.

Upps, added the missing register/fetch for the telegrams in 3.7.0-test32

**Comment by tp1de** [comment]
@proddy @MichaelDvP 

I implemented a prototype to manage EMS2.0 and EMS+ switchPrograms within Home Assistant using the HACS Scheduler compoment and Node-Red. This solution uses raw hex telegrams send / read by MQTT.  It works fine, but was quite tricky for my RC310 thermostat.

I wrote a document about my findings during this process and some analysis I have done on my original Bosch / Buderus gateway.
I made a proposal what needs to be done within EMS-ESP to support these entities in a better way.
Please have a look. I hope that I have been clear enough ....


 
[EMS switchPrograms.pdf](https://github.com/user-attachments/files/16760769/EMS.switchPrograms.pdf)

**Comment by MichaelDvP** [comment]
I've updated the test build https://github.com/MichaelDvP/EMS-ESP32/releases/tag/test
this should give a json structure for 
- api/thermostat/hc1/switchtime1
- api/thermostat/hc1/switchtime2
- api/thermostat/dhw/switchtime
- api/thermostat/dhw/circswitchtime

the switchtime1/2 depends on setting of hc1/switchprogmode.

I have not tested writing yet, the implementation is to send a single time in one command as json: `"value":{"day":"mo","time":"06:15","mode":"comfort","id":0}`. Writing a complete program needs 42 commands, but editing is much easier.

Also mqtt is not implemented. I think of an extra topic `thermostat_swichtimes` as nested or `thermostat_switchtime1_hc1`, `thermostat_circswitchtime_dhw`, etc as not nested.

**Comment by tp1de** [comment]
I installed your testbuild. 1st test (read) is ok. 
I would just recomment to keep the modes in English - today the are translated. This makes any logic flow to integrate in UI components  (e.g. HASS scheduler) language dependend.

Writing should be per day with up to 6 switchTimes. How do you indicate otherwise deleted switchTimes?

**Comment by MichaelDvP** [comment]
> I would just recomment to keep the modes in English

Ok
> Writing should be per day with up to 6 switchTimes. 

Does not work with RC20, RC30 and RC35 where day is coded and can be anywhere in the 42 data, Then i have to write all 42 at once.
> How do you indicate otherwise deleted switchTimes?

Reading: values are ignored, writing with `{"mode":"not_set",...}`
> .. and may I repeat my recommandation to name the entities switchprograms. T

We have translated word `switchtimes`, so i take this. Renaming and translating is a lot of work. If we have a complete concept and stick to it, we can rename. For now i dont make all the translations.

I've updated, setting should work now.

**Comment by tp1de** [comment]
> I've updated, setting should work now.

sorry Michael ... api does not work anymore - no response: 
message:   "no switchtime1 in thermostat"

**Comment by MichaelDvP** [comment]
Oh, sorry, did not mention, renamed the mqtt shortname to switchprog1, etc. as first step.

**Comment by tp1de** [comment]
Is there any reason that you haven't used the active program as well to select the switchprogram? (just switchprogmode)
That's the way it is done within Bosch / Buderus App. In this case you would just need switchprog and not 1 and 2.

What is the use of the no (index). It is counted per day?

**Comment by MichaelDvP** [comment]
> Is there any reason that you haven't used the active program as well to select the switchprogram?

Yes, we already have two progs for RC35 and i don't want to break this in first step. For RC300 it is better to use onlyactive program, also junkers, that have 6 programs. RC20 is difficult, we know the storage of the program, but not how to select it, it seems only to have predefined programs. RC35 have 11 programs, 2 of them user defined. The selected program is stored at the end of the telegram, so we dont know which program is selected on receiving the first parts of the telegram. Difficult and error prone. 
> What is the use of the no (index). It is counted per day?

Yes, for RC300 and Junkers  it counts 0..5, for RC20, RC30 andRC35 it counts 0..42. I rename to "no" to avoid confusion with id (hc).

**Comment by tp1de** [comment]
Thanks for your explanations - all the thermostat variants does it make quite complex.

Let me share my observations:
I recognized that all telegram types for switchPrograms are having read-requests every minute (even the inactive ones).
This increases the TX reads per 24hours from approx.  42.000 to 98.000. 
Can't this be done in a different way? (similar to original gateways - I tried to explain in the PDF)

In respect to the JSON structure and the no (index), I ask myself how to track changes in home automation systems and how to prepare the post commands. Without tracking changes all 42 switchTimes have to be send by single http posts. 
What kind of timedelay is needed inbetween?

Wouldn't it be easier by indicating in no the actual telegram position for all devices (for RC300/310 and Junkers too) and to send the full JSON in one http post?

**Comment by MichaelDvP** [comment]
I've updated the test build: https://github.com/MichaelDvP/EMS-ESP32/releases

- all UI changes from official dev33
- change names to switchprog (api/mqtt, all internal names, only long names are unchanged)
- fetch only active telegrams (except for RC30/35 program2 fetches prog1 and 2 because prog1 also contains other data)
- allow command as array: e.g. `{"day":"mo","no":2,"time":"12:00","mode":"off"}` for a single time, `[{"day":"mo","no":2,"time":"12:00","mode":"on"},{"day":"mo","no":3,"time":"12:30","mode":"off"}]` for two settings at once. With array command always the complete 84byte telegram is written.
- mqtt publish as topic: `thermostat_switchprog`, mqtt commands as usual to topic thermostat.
- In web UI customizations empty value shown for switchprogs, in devices it is shown as command (arrow) to allow writing,

I've tested only with my RC35. Because RC300 and Junkers are different there may be some issues.

**Comment by tp1de** [comment]
I tested on my RC310: (api gets first)

dhw entities are working for api reads. I just recognized that you kept the numbering per day - any reason why?
But the hc switchprog returns an empty array - no values. Please check.

**Comment by MichaelDvP** [comment]
>  I just recognized that you kept the numbering per day - any reason why?

I need them for RC20/30/35 and for deleting/changing only a single value. 
> But the hc switchprog returns an empty array - no values. Please check.

Please make a full log to see if i missed to fetch a telegram. But after start it needs some minutes, wait for hc active, first fetch to get the active program, second fetch to get the values.

**Comment by tp1de** [comment]
> **Edit:** found and fixed a typo for 1st absolute program, second absolute should work. Please update (in a few minutes, after github action has finished compiling)

After installing your last version reading seems to work for "level" and "absolute" and values are initialized in the right way after approx 90 to 100 seconds after start. Changing switchprogmode oder program is working as well.

>   I just recognized that you kept the numbering per day - any reason why?
I need them for RC20/30/35 and for deleting/changing only a single value.

I thought that for RC20-35 you will use the absolute position within telegram? 
(if yes, this could be done for EMS+ as well)

**Comment by MichaelDvP** [comment]
> (if yes, this could be done for EMS+ as well)

I think it makes no sense. For RC35 a value `"no":35, "day":"mo", "time":"00:30"` is valid, for RC300 each parameter is in range, but the combination is not valid because the day is asociated with number. So it's more clear to count the values within the day 0..5.
BTW: you can also use numbers for the json parameters. e.g. "day":0 instead of "day":"mo" or "time":135 instead of "time":"02:15"

**Comment by tp1de** [comment]
I checked within my running ioBroker instance and HA.

Within MQTT you have created one new element ems-esp/thermostat_switchprog containing all switchprograms.
This is nice, since I could create the HA entities - e.g. scheduler entities - from this data.
But how do changes would work - updating the whole structure?

The ioBroker adapter reads the device value list calling  .../api/thermostat. 
The switchprog entities are there but with an empty JSON: {}
Would be good to get the data content in as well to test if the adapter can be used without changes. (I hope so)

**Comment by MichaelDvP** [comment]
> Within MQTT you have created one new element ems-esp/thermostat_switchprog containing all switchprograms.
This is nice, since I could create the HA entities - e.g. scheduler entities - from this data.
But how do changes would work - updating the whole structure?

The topic is published in intervals of "other", fetch on the bus is every minute. Writing multiple points at once updates the buffer direct and sends the buffer to the bus. For single points the buffer update is actually on the next fetch. Mqtt publish on change is not yet implemented for the programs.

> The ioBroker adapter reads the device value list calling .../api/thermostat.
The switchprog entities are there but with an empty JSON: {}

Yes, i added this to tell your adapter that there is a value (and it's json). I think the response for RC300 with some hc is getting too big and blow the ram of chips without PSRAM.

>Would be good to get the data content in as well to test if the adapter can be used without changes. (I hope so)

I think the json data needs a special handling, so it also can be fetched seperately.
You can GET `/api/thermostat/hc1/switchprog/value` 

If you have a S32 or older gateway (ESP32 no PSRAM), try to get `api/thermostat/entities` Is it complete? (but afair you have only one hc).

**Comment by tp1de** [comment]
I am using an E32 V2.

api/thermostat/entities shows all data including switchprograms.

If you are concerned about memory allocation issues, than I need to introduce a separate call for each switchprogram to get the data. This would need adjustments to the ioBroker adapter. 
On the other side PSRAM (and free PSRAM) is known. Couldn't  you add json content depending on PSRAM value?

**Comment by MichaelDvP** [comment]
> Any chance to get a test version with data?

Sure, it's a test version, we can add and remove something to test. I've added to api/thermostat. For now for all, A Psram check is included, but commented out. Wait 10 min for compile done in github.

**Comment by tp1de** [comment]
Thanks Michael. 
To work with ioBroker object browser the json content needs to be stringified.
If your plan to enable it would be nice to change the coding. 
Otherwise I need to adjust the iobroker adapter.

**Comment by tp1de** [comment]
Nice job. Everything is working fine. No adjustments are needed for the iobroker adapter. Changing values are working as well.

**Comment by proddy** [comment]
cool! I'll take a look as well. Shall we squeeze this into 3.7.0 or release it after in a dev 3.7.1 when people have tested it? It's been 6+ months since the last stable release and BBQKees is still using 3.6.5 on his latest batch

**Comment by tp1de** [comment]
I will continue testing today. I do have 2 questions:

1. shall we avoid the backslashes within json stringify on .../api/thermostat call ? (just for nicer read)
I would keep them.

![grafik](https://github.com/user-attachments/assets/41a0496e-a289-4437-b3b9-1ddadfe88576)

2.  The switchprograms are not within HA entities - just the mqtt data in ems-esp/thermostat_switchprog
That's ok for me, but what is your opinion about it?
Anyhow HA does not know json entities - as far as I know. 
Text entities are limited to 255 characters. I was getting error log entries by decoding the switchprogram data from my km200 gateway in HA due to length and have limited the length therefore.
Does it makes sence to show the data in HA - even it can not be used easily?
I will create scheduler entities out of the mqtt data, similar to my todays solution with raw telegrams. (using node-red)
(I need to rewrite my nr flows)

**Comment by MichaelDvP** [comment]
I have updated to latest dev34-changes and refactored some parts of the switchprog code, hopefully without braking something. Please check.
> shall we avoid the backslashes within json stringify on .../api/thermostat call ? (just for nicer read)

The backslashes are needed for json-strings with inner quotationmarks. To avoid i have to send as json-array again. In `api/thermostat/switchprog` it's a json, in `api/thermostat/switchprog/value` a string not inside a json.
> The switchprograms are not within HA entities

I don't use HA and have no idea how to add this to HA. Also i think this will generate hundreds of mqtt-config messages. But we can try if you tell me what is needed for a HA config.

**Comment by MichaelDvP** [comment]
@proddy 
> Shall we squeeze this into 3.7.0 or release it after in a dev 3.7.1

There is also Junkers switchprog that needs testing. I think i've also broke some modbus assignes, can't get the scripts running on first try. For the switchprog there is also no idea how to integrate in web-UI, now it's just like a command to change values, but no display for the program. BTW: RC20 switchprog should also work with telegram 0x8F, RC20_N not, don't know what telegram is used., maybe same as RC30/35? (0x3F).

If you want to release soon, than better merge this to 3.7.1.

**Comment by tp1de** [comment]
> I have updated to latest dev34-changes and refactored some parts of the switchprog code, hopefully without braking something. Please check.

Works still fine.

Json stringify is fine for me and reflects the json standards.

For HA I think creating entities from EMS-ESP will work only with discovery and putting the switchprogram json into the attributes.
Since the HASS scheduler is using an own entity storage these schedules have to be created by using the scheduler service calls.
This can be done by reading the mqtt data. I do not see any need to go for ha discovery in parallel, unless @proddy has different ideas.

**Comment by tp1de** [comment]
@MichaelDvP 
I tested on my second installation: my Dutch holidayhouse with Moduline 2050 EMS 2.0 thermostat.

This thermostat does not have any switchprogmode and just one program. 
It is permanent "level" and prog1 for hc1.

In this case no switchprogram is discovered. I think this might be caused by the non-existing switchprogmode.

There is no value but by calling per api it is shown, but without index. (it should not exist)

![grafik](https://github.com/user-attachments/assets/abbaa6d1-8b16-438c-9a44-801f64501d9c)

**Comment by tp1de** [comment]
The Moduline 2050 has  deviceID: 0x18  and productId: 157 / version 43.03
switchprogmode is not shown due to missing index. This is ok, but then you should use "level". 
For program you show both prog1 and prog2, but prog 2 does not exist.

**Comment by MichaelDvP** [comment]
I've removed program and switchprogmode from RC100 compatible thermostats and switchprog-default to first telegram.

**Comment by tp1de** [comment]
> I've removed program and switchprogmode from RC100 compatible thermostats and switchprog-default to first telegram.

Yes it works now.

**Comment by tp1de** [comment]
On my RC310 system I was testing my scheduler scripts now adjusted to the new scheduler.
I am getting error messages (not always) when I change the schedule/switchprograms.
I am using 2 schedules: one for workdays (mo-fr) and one for weekend (sa-su). 
The changes are send per schedule when saved using days with numbers. E.g.

2024-09-07 14:57:58.769 I 69: [command] Called command thermostat/switchprog (Programm Schaltzeit) with value [{"day":0,"time":"06:15","mode":"comfort","no":0},{"day":0,"time":"21:30","mode":"eco","no":1},{"day":1,"time":"06:15","mode":"comfort","no":0},{"day":1,"time":"21:30","mode":"eco","no" :1 .....
2024-09-07 14:57:59.295 E 70: [emsesp] Last Tx write rejected by host
2024-09-07 14:58:01.743 E 71: [emsesp] Last Tx write rejected by host
2024-09-07 14:58:02.545 E 72: [emsesp] Last Tx write rejected by host

At the end the changes are done but the mqtt entity ems-esp/thermostat/switch_prog disapears for approx. 1-2 minutes.

What are the constraints of sending changes?

**Comment by MichaelDvP** [comment]
> Last Tx write rejected by host

This is from the bus, that does not accept the write. I send the 84 bytes in 4 parts. I've reordert to start with offset 0, maybe this helps. Please make a full log and trigger this errormessage to see what part is rejected and if the retry is accepted.
> Looks like a memory problem.

I forgot to check valid no-values for RC300, was the `no` greater 5 for sunday?

**Comment by tp1de** [comment]
> I forgot to check valid no-values for RC300, was the `no` greater 5 for sunday?

days are 0 to 4 for mo to fr and 5 and 6 for sa and su.

All changes for either 5 or 2 days are done correctly - just the error messages are in log.
So retry is working. I could not produce any crash anymore. 

I was using a raw mode for the last weeks and needed to send 4 telegrams as well.
I recognized that I need a wait time inbetween the telegrams to avoid errors. 
After some testing I took 1 second waiting between the telegrams for 100% writing accuracy.

**Comment by MichaelDvP** [comment]
Please try with full log if you can resend and trigger the error-log-message. I like to see if the reject was repeated on one part or on each new part like: 1part -> 2.part -> reject ->retry2.part ->3.part -> reject -> retry3.part -> 4.part ->reject -> retry

**Comment by tp1de** [comment]
Full log for changing weekend:

[log.txt](https://github.com/user-attachments/files/16919171/log.txt)

**Comment by tp1de** [comment]
I used my syslog tool to filter 2C3 telegrams (0x01C3 raw) while changing 2 days (weekend: sa and su)):

9/7/2024, 7:46:46 PM | 10 00 FF 3C 01 C3 03 18 01 58 03 FF 03 FF 03 FF 03 FF 03 18 01 58 03 FF 03 FF 03 FF 03 FF 11
9/7/2024, 7:46:45 PM | 0B 10 FF 42 01 C3 03 FF 03 FF 03 FF 03 18 01 58 03 FF 03 FF 03 FF 03 FF 0B
9/7/2024, 7:46:43 PM | 10 00 FF 24 01 C3 03 1C 01 56 03 FF 03 FF 03 FF 03 FF 03 1C 01 56 03 FF 03 FF 03 FF 03 FF E4
9/7/2024, 7:46:43 PM | 0B 10 FF 2C 01 C3 03 FF 03 FF 03 1C 01 56 03 FF 03 FF 03 FF 03 FF 03 18 01 58 03 FF 4B
9/7/2024, 7:46:42 PM | 10 00 FF 24 01 C3 03 1C 01 56 03 FF 03 FF 03 FF 03 FF 96
9/7/2024, 7:46:42 PM | 10 00 FF 0C 01 C3 03 1C 01 56 03 FF 03 FF 03 FF 03 FF 03 1C 01 56 03 FF 03 FF 03 FF 03 FF 96
9/7/2024, 7:46:42 PM | 0B 10 FF 16 01 C3 03 FF 03 1C 01 56 03 FF 03 FF 03 FF 03 FF 03 1C 01 56 03 FF 03 FF 56
9/7/2024, 7:46:41 PM | 10 00 FF 00 01 C3 03 1C 01 56 03 FF 03 FF 03 FF 03 FF 03 1C 01 56 03 FF 03 FF 03 FF 03 FF 81
9/7/2024, 7:46:41 PM | 0B 10 FF 00 01 C3 03 1C 01 56 03 FF 03 FF 03 FF 03 FF 03 1C 01 56 03 FF 03 FF 03 FF F1

Could it be that your telegrams are too short?

**Comment by MichaelDvP** [comment]
Ok, myRC35 don't care about splitting, but i don't want to split odd withing a switchpoint, so i used 22/22/22/18, I've changed now to 24/24/24/12, maybe this is better for RC300. 

The full log shows 3 repeats for the part with offset 2C, 3 times as incomplete, i've marked the missing/wrong bytes :
```
0B 10 FF 2C 01 C3 03 FF 03 FF ** 18 01 56 03 FF 03 FF 03 FF 03 FF 03 1A 01 58 03 FF 12
0B 10 FF 2C 01 C3 03 FF 03 FF 03 18 01 56 03 FF 03 FF 03 FF 03 FF 03 1A 01 58 03 FF *02 90*
0B 10 FF 2C ** C3 03 FF 03 FF 03 18 01 56 03 FF 03 FF 03 FF 03 FF 03 1A 01 58 03 FF 12

0B 10 FF 2C 01 C3 03 FF 03 FF 03 18 01 56 03 FF 03 FF 03 FF 03 FF 03 1A 01 58 03 FF 12
```
The last try is corrcet and works., the other parts work on first try. Don't know why.

**Comment by tp1de** [comment]
I still get some errors, but much less than before. 
Is it a timing issue between the telegrams?

**Comment by MichaelDvP** [comment]
Please add a log, are the broadcast responses from thermostat now on the same offset as sended? 
Maybe we need split to single days, e.g. 12 bytes by telegram? 
> Is it a timing issue between the telegrams?

I don't think so, Emsesp sends only when the master polls, there is no other timing possible.

**Comment by tp1de** [comment]
Could you please check your code another time. The telegrams do not look right for me:

9/8/2024, 12:44:26 PM | 0B 10 FF 42 01 C3 01 FF 01 FF 01 FF 03 16 01 58 01 FF 01 FF 01 FF 01 FF 3B
9/8/2024, 12:44:26 PM | 10 00 FF 3C 01 C3 03 16 01 58 01 FF 01 FF 01 FF 01 FF 4F
9/8/2024, 12:44:26 PM | 10 00 FF 24 01 C3 03 17 01 56 01 FF 01 FF 01 FF 01 FF 03 17 01 56 01 FF 01 FF 01 FF 01 FF 12
9/8/2024, 12:44:25 PM | 10 00 FF 18 01 C3 03 17 01 56 01 FF 01 FF 01 FF 01 FF 03 17 01 56 01 FF 01 FF 01 FF 01 FF 59
9/8/2024, 12:44:25 PM | 0B 10 FF 2C 01 C3 01 FF 01 FF 03 17 01 56 01 FF 01 FF 01 FF 01 FF 03 16 01 58 01 FF 75
9/8/2024, 12:44:24 PM | 10 00 FF 00 01 C3 03 17 01 56 01 FF 01 FF 01 FF 01 FF 03 17 01 56 01 FF 01 FF 01 FF 01 FF 77
9/8/2024, 12:44:24 PM | 0B 10 FF 16 01 C3 01 FF 03 17 01 56 01 FF 01 FF 01 FF 01 FF 03 17 01 56 01 FF 01 FF A4
9/8/2024, 12:44:23 PM | 0B 10 FF 00 01 C3 03 17 01 56 01 FF 01 FF 01 FF 01 FF 03 17 01 56 01 FF 01 FF 01 FF 41

I tested with sending raw telegrams - each telegram 2 days = 24 bytes of data. Send by mqtt.
Without any wait time inbetween I get the same errors: [emsesp] Last Tx write rejected by host
Wait time between commands > 500msec is ok and without errors.

**Comment by MichaelDvP** [comment]
I think you have picked a intermediate buggy version that was only a few minutes online. Murphys law.

**Comment by tp1de** [comment]
These are the telegrams when I send raw using your last version:

9/8/2024, 7:13:35 PM | 10 00 FF 3C 01 C3 03 16 01 56 01 FF 01 FF 01 FF 01 FF 03 16 01 56 01 FF 01 FF 01 FF 01 FF 87
9/8/2024, 7:13:35 PM | 10 0B FF 3C 01 C3 03 16 01 56 01 FF 01 FF 01 FF 01 FF 03 16 01 56 01 FF 01 FF 01 FF 01 FF F8
9/8/2024, 7:13:35 PM | 10 00 FF 30 01 C3 03 16 01 56 01 FF 01 FF 01 FF 01 FF 39
9/8/2024, 7:13:35 PM | 0B 10 FF 3C 01 C3 03 16 01 56 01 FF 01 FF 01 FF 01 FF 03 16 01 56 01 FF 01 FF 01 FF 01 FF 10
9/8/2024, 7:13:34 PM | 10 0B FF 30 01 C3 03 16 01 56 01 FF 01 FF 01 FF 01 FF 23
9/8/2024, 7:13:34 PM | 10 00 FF 18 01 C3 03 16 01 56 01 FF 01 FF 01 FF 01 FF 03 16 01 56 01 FF 01 FF 01 FF 01 FF E2
9/8/2024, 7:13:34 PM | 0B 10 FF 30 01 C3 03 16 01 56 01 FF 01 FF 01 FF 01 FF 13
9/8/2024, 7:13:32 PM | 10 0B FF 18 01 C3 03 16 01 56 01 FF 01 FF 01 FF 01 FF 03 16 01 56 01 FF 01 FF 01 FF 01 FF 9D
9/8/2024, 7:13:32 PM | 0B 10 FF 18 01 C3 03 16 01 56 01 FF 01 FF 01 FF 01 FF 03 16 01 56 01 FF 01 FF 01 FF 01 FF 75
9/8/2024, 7:13:31 PM | 10 00 FF 00 01 C3 03 16 01 56 01 FF 01 FF 01 FF 01 FF 03 16 01 56 01 FF 01 FF 01 FF 01 FF CC
9/8/2024, 7:13:31 PM | 10 0B FF 00 01 C3 03 16 01 56 01 FF 01 FF 01 FF 01 FF 03 16 01 56 01 FF 01 FF 01 FF 01 FF B3
9/8/2024, 7:13:31 PM | 0B 10 FF 00 01 C3 03 16 01 56 01 FF 01 FF 01 FF 01 FF 03 16 01 56 01 FF 01 FF 01 FF 01 FF 5B

and these are the telgrams when I send to json:

9/8/2024, 7:20:35 PM | 10 00 FF 3C 01 C3 03 16 01 56 01 FF 01 FF 01 FF 01 FF 03 16 01 56 01 FF 01 FF 01 FF 01 FF 87
9/8/2024, 7:20:34 PM | 10 00 FF 3C 01 C3 03 16 01 56 01 FF 01 FF 01 FF 01 FF D1
9/8/2024, 7:20:34 PM | 10 00 FF 24 01 C3 03 16 01 56 01 FF 01 FF 01 FF 01 FF 03 16 01 56 01 FF 01 FF 01 FF 01 FF A9
9/8/2024, 7:20:34 PM | 0B 10 FF 42 01 C3 01 FF 01 FF 01 FF 03 16 01 56 01 FF 01 FF 01 FF 01 FF A5
9/8/2024, 7:20:33 PM | 0B 10 FF 2C 01 C3 01 FF 01 FF 03 16 01 56 01 FF 01 FF 01 FF 01 FF 03 16 01 56 01 FF 15
9/8/2024, 7:20:33 PM | 10 00 FF 24 01 C3 03 16 01 56 01 FF 01 FF 01 FF 01 FF 18
9/8/2024, 7:20:33 PM | 10 00 FF 0C 01 C3 03 16 01 56 01 FF 01 FF 01 FF 01 FF 03 16 01 56 01 FF 01 FF 01 FF 01 FF DB
9/8/2024, 7:20:32 PM | 10 00 FF 00 01 C3 03 16 01 56 01 FF 01 FF 01 FF 01 FF 03 16 01 56 01 FF 01 FF 01 FF 01 FF CC
9/8/2024, 7:20:32 PM | 0B 10 FF 16 01 C3 01 FF 03 16 01 56 01 FF 01 FF 01 FF 01 FF 03 16 01 56 01 FF 01 FF 9D
9/8/2024, 7:20:32 PM | 0B 10 FF 00 01 C3 03 16 01 56 01 FF 01 FF 01 FF 01 FF 03 16 01 56 01 FF 01 FF 01 FF A5

**Comment by MichaelDvP** [comment]
> Could you please check your code another time. The telegrams do not look right for me:

Checked again. Your log shows:
`0B 10 FF 00 01 C3 03 16 01 56 01 FF 01 FF 01 FF 01 FF 03 16 01 56 01 FF 01 FF 01 FF A5`
22 bytes long and next telegram with offset 22, also 22 bytes long. 
`0B 10 FF 16 01 C3 01 FF 03 16 01 56 01 FF 01 FF 01 FF 01 FF 03 16 01 56 01 FF 01 FF 9D`

And now look at this commit:  https://github.com/MichaelDvP/EMS-ESP32/commit/1be2eea3c2ab73835c6f20a4404b4852da38516f

This is not the same software.

**Comment by tp1de** [comment]
I took from here 
https://github.com/MichaelDvP/EMS-ESP32/releases/tag/test

**Comment by MichaelDvP** [comment]
I've made a commit only to update version, so you can check if you have the latest version installed. 
For me it gives telegrams with 24 bytes and no errors:
```
2024-09-08 21:25:58.972 INFO 52: [thermostat] Reading jsonArray
2024-09-08 21:25:58.972 INFO 53: [thermostat] id: 0, day: mo, mode: on, time: 05:50
2024-09-08 21:25:58.972 INFO 54: [thermostat] id: 1, day: mo, mode: off, time: 21:30
2024-09-08 21:25:58.972 INFO 55: [thermostat] id: 2, day: tu, mode: on, time: 05:50
2024-09-08 21:25:58.972 INFO 56: [thermostat] id: 3, day: tu, mode: off, time: 21:30
2024-09-08 21:25:58.972 INFO 57: [thermostat] id: 4, day: we, mode: on, time: 05:50
2024-09-08 21:25:58.972 INFO 58: [thermostat] id: 5, day: we, mode: off, time: 21:30
2024-09-08 21:25:58.972 INFO 59: [thermostat] id: 6, day: th, mode: on, time: 05:50
2024-09-08 21:25:58.972 INFO 60: [thermostat] id: 7, day: th, mode: off, time: 21:30
2024-09-08 21:25:58.972 INFO 61: [thermostat] id: 8, day: fr, mode: on, time: 05:50
2024-09-08 21:25:58.972 INFO 62: [thermostat] id: 9, day: fr, mode: off, time: 21:30
2024-09-08 21:25:58.972 INFO 63: [thermostat] id: 10, day: sa, mode: on, time: 06:20
2024-09-08 21:25:58.972 INFO 64: [thermostat] id: 11, day: sa, mode: off, time: 21:30
2024-09-08 21:25:58.972 INFO 65: [thermostat] id: 12, day: su, mode: on, time: 07:00
2024-09-08 21:25:58.972 INFO 66: [thermostat] id: 13, day: su, mode: off, time: 21:30
2024-09-08 21:25:58.972 INFO 67: [command] Called command thermostat/switchprog (program switchtime) with value [{"no":0,"day":"mo","time":"05:50","mode":"on"},{"no":1,"day":"mo","time":"21:30","mode":"off"},{"no":2,"day":"tu","time":"05:50","mode":"on"},{"no":3,"day":"tu","time":"21:30","mode":"o
2024-09-08 21:25:59.378 TRACE 72: [emsesp] Me(0x4A) -W-> thermostat(0x10), RC35Timer(0x49), data: 01 23 00 81 21 23 20 81 41 23 40 81 61 23 60 81 81 23 80 81 A1 26 A0 81
2024-09-08 21:26:00.266 TRACE 81: [emsesp] Me(0x4A) -W-> thermostat(0x10), RC35Timer(0x49), data: C1 2A C0 81 FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF (offset 24)
2024-09-08 21:26:00.891 TRACE 86: [emsesp] Me(0x4A) -W-> thermostat(0x10), RC35Timer(0x49), data: FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF (offset 48)
2024-09-08 21:26:01.383 TRACE 89: [emsesp] Me(0x4A) -W-> thermostat(0x10), RC35Timer(0x49), data: FF FF FF FF FF FF FF FF FF 90 E7 90 (offset 72)
```

**Comment by tp1de** [comment]
Sorry Michael,  this is difficult for me to understand. 
I installed your latest version test35 . Result in respect to length and offset is unchanged.

**Comment by tp1de** [comment]
My gateway crashed this morning again and lost wifi connection. 
I am not at home and can't restart. 
Something is instable in current firmware.

**Comment by tp1de** [comment]
@MichaelDvP 
I installed the node-red scripts for switchprograms as well on my second installation and compared the telegrams:

1. Moduline 2050
Telegrams have the right length and offset and thermostat replies are fine as well as far as I can judge.
Why are errors shown in log?
[moduline 2050.txt](https://github.com/user-attachments/files/16934674/moduline.2050.txt)

2. RC310
Telegram length and offsets are different -why? Do they depend on thermostat type?
[RC310.txt](https://github.com/user-attachments/files/16934690/RC310.txt)

In both cases the post commands are executed and the switchprograms are changed.

**Comment by MichaelDvP** [comment]
Seems the moduline does not like the 24bytes telegrams, no error for the 12byte part. Let's try to send 12byte telegrams.
I'll push test36 version.

RC310 is still test34 with 22 byte. I think you need to clear browser cache before updating, There were some changes in update-code of dev34 that i've merged to keep up to date with the dev.

**Comment by tp1de** [comment]
> Seems the moduline does not like the 24bytes telegrams, no error for the 12byte part. Let's try to send 12byte telegrams.
> I'll push test36 version.

Does not make it better:
[moduline 2050 2.txt](https://github.com/user-attachments/files/16945587/moduline.2050.2.txt)

**Comment by tp1de** [comment]
Doing changes on moduline 2050 for mo-fr sends this telegrams:

9/10/2024, 3:51:35 PM | 18 00 FF 30 01 C3 03 18 01 1A 03 FF 01 FF 03 FF 01 FF 3D
9/10/2024, 3:51:34 PM | 18 00 FF 18 01 C3 03 18 01 1A 03 FF 01 FF 03 FF 01 FF 03 18 01 1A 03 FF 01 FF 03 FF 01 FF 0D
9/10/2024, 3:51:34 PM | 18 00 FF 00 01 C3 03 18 01 1A 03 FF 01 FF 03 FF 01 FF 03 18 01 1A 03 FF 01 FF 03 FF 01 FF 23

**Comment by tp1de** [comment]
Doing changes on RC310 using the cloud app and the gateway for monday and then copying to all other days sends this telegrams:

Copy Monday to other days:
9/10/2024, 3:38:25 PM	10 00 FF 3C 01 C3 03 16 01 56 03 FF 03 FF 03 FF 03 FF 03 16 01 56 03 FF 03 FF 03 FF 03 FF 00
9/10/2024, 3:38:25 PM	10 00 FF 24 01 C3 03 16 01 56 03 FF 03 FF 03 FF 03 FF 03 16 01 56 03 FF 03 FF 03 FF 03 FF 2E
9/10/2024, 3:38:24 PM	10 00 FF 0C 01 C3 03 16 01 56 03 FF 03 FF 03 FF 03 FF 03 16 01 56 03 FF 03 FF 03 FF 03 FF 5C
9/10/2024, 3:38:24 PM	48 10 FF 3E 01 C3 01 56 FF FF FF FF FF FF FF FF 03 16 01 56 FF FF FF FF FF FF FF FF E4
9/10/2024, 3:38:23 PM	48 10 FF 25 01 C3 16 01 56 FF FF FF FF FF FF FF FF 03 16 01 56 FF FF FF FF FF FF FF FF 03 16 EE
9/10/2024, 3:38:23 PM	48 10 FF 0C 01 C3 03 16 01 56 FF FF FF FF FF FF FF FF 03 16 01 56 FF FF FF FF FF FF FF FF 03 A9

Monday changes:
9/10/2024, 3:37:23 PM	10 00 FF 00 01 C3 03 16 01 56 03 FF 03 FF 03 FF 03 FF FF
9/10/2024, 3:37:22 PM	48 10 FF 00 01 C3 03 16 01 56 FF FF FF FF FF FF FF FF 71

**Comment by tp1de** [comment]
I tried all TX modes. They are all worse compared to EMS. Moduline is a EMS 2.0 thermostat.

**Comment by MichaelDvP** [comment]
repeated wrong master echos that shows different receive from sending (in this case an extra byte after crc):
`Incomplete Rx: 0B 18 FF 24 01 C3 03 18 01 58 01 FF 01 FF 01 FF 01 FF 14 89`
`Last Tx Write operation failed after 3 retries. Ignoring request: 0B 18 FF 24 01 C3 03 18 01 58 01 FF 01 FF 01 FF 01 FF`
is wrong timing (tx-mode) or bad cabeling.

**Comment by tp1de** [comment]
> is wrong timing (tx-mode) or bad cabeling.

I agree that this is a timing problem. I tested with sending raw telegrams per days - so 7 telegrams without delay:
The result could be

(1) [moduline 2050 3.txt](https://github.com/user-attachments/files/16979261/moduline.2050.3.txt)
 or 
(2) [moduline 2050 4.txt](https://github.com/user-attachments/files/16979266/moduline.2050.4.txt)

In (1) there are a lot of errors and more other telegrams send by thermostat
In (2) there are no errors and just very few other telegrams send by thermostat

And it can be recognized that the send commands are processed in reverse order - the last one first.

I do not think that this a cabeling problem. (ems-esp connected by service jack and thermostat with 5 m cable) 
Could it be that the thermostat is processing rather slow and blocks other requests when occupied?
Or could it be a controller problem?  (BC10 - deviceID : "0x09" - productID : 190)

**Comment by MichaelDvP** [comment]
Does your boiler use buderus or ht3 protocoll? You'll find it in support info.
The bad telegrams always ends with an extra byte after the CRC, sometimes 98, sometimes 89. In ht3 this could be sending from thermostat (0x18 | 0x80) and  controller (0x09 | 0x80). But the controler normally 
does not send. 
Or the master misses the the break, timeouts, and this is the (buderus-protocol) poll to next device.
> And it can be recognized that the send commands are processed in reverse order - the last one first.

Yes, all tx messages are queued and waiting for the master poll. The periodic reads go to end of queue, writes go to top of queue to be send first. If you queue many writes each new goes to top and is sended out first.

**Comment by tp1de** [comment]
The boiler is a Netfit Trendline II from Bosch. It uses EMS.

**Comment by proddy** [comment]
I have the same boiler..the Dutch Nefit variant, but a Trendline II CW5. Can I help test something?

**Comment by MichaelDvP** [comment]
I've already figured out that it's buderus protocoll. The `Incomplete Rx` is sent before the ht3-bit stripped.
So the extra byte is not from thermostat, but the thermostat also receives it and sends NACK back.

@proddy Do you have seen such effects on writing something to thermostat (i know you have RC20, but it's also a write to device 0x18).

**Comment by tp1de** [comment]
The moduline 2050 supports the ems protocol and as well open therm.
Could this be the reason for special protocol handling?

**Comment by MichaelDvP** [comment]
@proddy afaik the RC20 has timer on telegram 0x8F, does this work with my testbuild? 
@tp1de Do you see the same appended byte on short telegrams like seltemp? And if you write only a single switchtime in the program (not asarray `[{...},{..}]`, only as jsonObject `{...}`)?

**Comment by tp1de** [comment]
> Do you see the same appended byte on short telegrams like seltemp?

No. And as explained comparing the two logs moduline 2050 3 and 4 you can see, that under certain conditions the 7 switchtimes writes work without any appended byte as well. It is a matter of other telegrams send inbetween. (total traffic during processing the changes)

It is a guess:  Does the appended byte indicates that the telegram should be resend or that the telegrams is stored and not written immedeatly but delayed ....??

**Comment by proddy** [comment]
> @proddy afaik the RC20 has timer on telegram 0x8F, does this work with my testbuild? 

I tested your build, saw the `hp1 program switchtime` command in the WebUI, entered the value `[{"no":0,"day":"mo","time":"05:50","mode":"on"}]` to see what happens and I see the telegrams sent, no errors. I checked on the physical RC20, and all the clock programs have been erased. So something is working! Do you have any test data I can use?

**Comment by MichaelDvP** [comment]
>  all the clock programs have been erased. 

Oh,bad. have you sent the command direct after starting before the telegram was read?. I will add a check for this.
> Do you have any test data I can use?

Only for RC300,  i have no telegram data for RC20, just a notice about timer telegram typeid. For RC20_N i don't know the timer-telegram(s).

@tp1de 
> Does the appended byte indicate that the telegram should be resend or that the telegrams is stored and not written immediately but delayed ....??

No, the thermostat also gets the wrong bytes appended and answers with a "not-ack"-byte. It's something from the busmaster. 
On ems bus the master provides the voltage and sends to the client with voltage modulation. The clients (and emsesp) only reads voltage modulation(rx), but sends as current modulation. Only the master can read current modulation and repeats byte by byte the sending of the client, so other clients can receive it. Here emsesp sends via current modulation the correct telegram and the master repeats with an extra byte. emsesp and the thermostat receives now a wrong telegram and the thermostats send (as current modulation) the NACK which is repeated (as voltage modulation) by the master and received from emsesp.

**Comment by tp1de** [comment]
I do have difficulties to interprete the consequences of errors when writing the switchtimes.
The log shows:

 _Last Tx Write operation failed after 3 retries. Ignoring request:0B 18 FF 00 01 C3 03 16 01 58 03 FF 03 FF 03 FF 03 FF_

I would expect that changes are not written. But this not the case.
All changes are updated correctly.

See log:
[moduline 2050 5.txt](https://github.com/user-attachments/files/17015439/moduline.2050.5.txt)

**Comment by Vippis2000** [comment]
Will this be integrated in a future dev. release? Right now I don't know what to do :)

**Comment by tp1de** [comment]
This is not the format from the testbuild from @MichaelDvP .
Are you using a  Buderus gateway? 

The RC310 supports 6 switchtimes per day.
The times must be aligned to 15 minutes.

**Comment by proddy** [comment]
@tp1de what's the latest status of supporting programmable times into the RC3x?

**Comment by tp1de** [comment]
> @tp1de what's the latest status of supporting programmable times into the RC3x?

@proddy  I tested again with latest test version from @MichaelDvP. It is still the same version:

RC310:  OK from my point of view
- Reading switchPrograms is working correctly
- On Update sometimes error messages in log "Last Tx write rejected by host" 
- all changes are correctly written

Moduline2050: ????
- Reading switchProgram is working correctly 
- On Update a lot of error messages in log "Last Tx write rejected by host"  and "Last Tx Write operation failed after 3 retries. Ignoring request: ..." for each sub-telegram send.
- I would expect that changes are not updated. But changes are correctly written to thermostat.

How shall we proceed?

**Comment by proddy** [comment]
Is the Home Assistant component also working, that Scheduler one from HACS. Then we can do an article on it.

**Comment by tp1de** [comment]
Within ioBroker the JSON-content of the switchProgram is available as JSON object - similar to km200. No graphical interface available, unless users create their own ones. 

Within HA there is no autodiscover entity for switchPrograms existing. So only the mqtt object thermostat_switchprog exists on the mqtt broker. 
To use scheduler from HACS the 2 components (scheduler and card) must be installed using hacs.
Then a program - in my case node-red flow - is needed to generate the scheduler components. The coding must be adjusted by users. Each Scheduler schedule is linked to days being active. If different schedules for workdays and weekends  are needed then code must be adjusted to generate 2 schedules per heating circuits. (or more)
Next to it helper objects need to be created manually for having this working. Without knowledge of node-red and javascript coding it would not be easy to use. 

I need to think about if there is a more easy way of using HACS Scheduler.

**Comment by tp1de** [comment]
@proddy @MichaelDvP 

I rewrote my node-red, made some documentation and uploaded everything:
https://github.com/tp1de/home-assistant-node-red-switchprog

Please have a look if it is to complicated for end-users and please give some feedback.
Shall we still use Michaels test-build or shall we release the switchPrograms into dev?

P.S.: I uploaded my NR flows for heatdemand calculation and reading the km200 gateway as well into Github.

**Comment by proddy** [comment]
I'll review, nice piece of work Thomas. I've been exploring creating a web scheduler for EMS-ESP, similar to [scheduler-component](https://github.com/nielsfaber/scheduler-component) and have a prototype ready. Maybe this is good use case for thermostat programs too.

**Comment by tp1de** [comment]
> I'll review, nice piece of work Thomas. I've been exploring creating a web scheduler for EMS-ESP, similar to [scheduler-component](https://github.com/nielsfaber/scheduler-component) and have a prototype ready. Maybe this is good use case for thermostat programs too. 

Good to know. The scheduler-component is not really needed just an updated version on the scheduler-card to update the time schedule.
If you have a scheduler to be tested .... I am happy to test.

**Comment by tp1de** [comment]
@MichaelDvP 
I observed errors while changund program mode from "Level" to "Absolute".
![grafik](https://github.com/user-attachments/assets/575b51e5-e8f9-4147-b377-cea9226cedc7)

switchprogmode is changed correctly in ems-esp dashboard and in HA while not recognized within code.

**Comment by MichaelDvP** [comment]
Was this the first test with absolute setting, or is something broken that works before?
Maybe it's the `"temp":19.5`, i had the `"temp" as int, changed now to float in new testbuild.

**Comment by tp1de** [comment]
@MichaelDvP 
I recognized that while changing program mode it takes 2 minutes (2 cycles) until mqtt content is changed.
1st mqtt message send is still with old content and format.

Within my node-red flow I read program and mode and build the structure accordingly. 
This might be too early and causing the error.

Why does it take 2 cycles to recognize the changes within ems-esp/thermostat_switchprog? 
Best case would be while changing either program or mode to send a new switchprog message immediately.

**Comment by MichaelDvP** [comment]
I think it's fixed now (bin will be generated in about 10 minutes), reading the new program should be a bit faster, but still we have to wait for next telegram. I use only one buffer for the program to save memory.

**Comment by tp1de** [comment]
I do not see any changes. 
It still takes 2 cycles until mode changes are seen and non int temp values still do not work while updating.

**Comment by tp1de** [comment]
I do not see any changes in respect to timing. It still takes 2 cycles until mode changes are seen.
Updating non int values on "Absolute" mode is working now.

**Comment by tp1de** [comment]
I think I found what causes the problem:

While updating thermostat_switchprograms I send data for 5 days (workdays), 2 days (weekend) or 7 days for (daily).
The code will transform this to telegram send. There are regulare messages: Last Tx write rejected by host
The result will be that not all days are updated. Since I take Monday for workdays and Saturday for weekend this results in the situation described.

I still have my node-red workflow requesting and sending raw telegrams. While sending I send telegrams per day.
Doing this without any delay will result in the same result (errors). I tested different delays. Having more then 100 msec inbetween it is working reliable. I decided to use even 1000 msec to be on the save side.

I tested the behavior while changing the switchProgram with the MyBuderus app using the km200 gateway. While copying the values from one day to all other 6 days there are 3 telegrams send (2 days within one telegram). The delay is approx 400 msec between the telegrams send.

**Comment by MichaelDvP** [comment]
I've updated to a logic to readback the part after bus write before rend the next part. Maybe this gives the delay needed. 
Also writing the absolute telegram should work now. After the mode change i trigger a read out of order of normal fetch interval, i think faster is not possible, or we have to fetch and buffer all 4 timer-telegrams.

**Comment by tp1de** [comment]
> After the mode change i trigger a read out of order of normal fetch interval, i think faster is not possible, or we have to fetch and buffer all 4 timer-telegrams.

Yes it it is working for mode changes (next cycle) . Still not for changes of program (2 cycles)

> I've updated to a logic to readback the part after bus write before rend the next part. Maybe this gives the delay needed.

Updates still do not work reliable. Approx 30-40% without errors (Last Tx write rejected by host) and the rest have 1 to 3 errors while updating 5 days. 
I tested to send the update twice with 5 seconds inbetween - this is updating then all days but will this be 100% ?

**Comment by tp1de** [comment]
I tested again with my raw telegram version.

- using a telegram per day a time delay > 100 msec is sufficient not to get tx write errors
-  using telegram length of 2 days a delay of > 300 msec is needed

There is no need to go too low to this limits. Can't you just intoduce a simple delay inbetween the telegrams send?

**Comment by MichaelDvP** [comment]
Updated with 400ms delay between sending telegrams with same type-id. Please check.

**Comment by MichaelDvP** [comment]
> I get for some tx writes error messages in log like "Last Tx write host reply: 0xB0" with different hex values.
What is the meaning for this log entries?

The reply should be OK (01) or REJECTED (04), other values are not known. 
I show them, but think this is not a reply and the master timeouts and just send a new poll to a different device. But maybe there is another reply code we don't know.

**Comment by tp1de** [comment]
I did extensive test with more then 100 updates on switchprograms on my heating system.
At the end I can't avoid errors or read requests (Last Tx Read operation failed after 3 retries. Ignoring request: 0B 10 FF 00 02 09 0C) or on write requests (Last Tx write rejected by host).
Testing with raw telegrams these errors less often when increasing the delay inbetween the commands.
But even with 2 seconds inbetween there are errors on maybe 5-10% of all cases.
Your updated testbuild with 400ms delay produces less errors but it is not error free.

At the end the update of switchprograms is not reliable since there is a chance of having 1-2 days not updated.

I do not see any difference between km200 ems-esp initiated commands towards thermostat. Maybe you can see the difference?
[watch.txt](https://github.com/user-attachments/files/17879150/watch.txt)

The gateway intializes  non used switchTimes with "FF FF" while we are using "03 FF". I tested both versions on my raw version. This does not have any effect on the tx  write errors.

I have no idea what to do else ......

**Comment by MichaelDvP** [comment]
Difficult, there are two commands withing a few ms, Why? So i can't see what was rejected, what executet.
Differences to gateway:
- gateway delay is around one second
- gateway sends a small block first (12 bytes), and larger after, ems-esp starts with 24 bytes.
I'll update to 1 sec delay and 12 bytes first, maybe this works better. (-test.10)
Please try to change one time per block, and log states before and after write, s we can see if a `reject` ignores the values or writes them. There is a reject, but thermostat publishes the first part, what normally only happen after a successful write.

**Comment by tp1de** [comment]
> * gateway sends a small block first (12 bytes), and larger after, ems-esp starts with 24 bytes.

This because of MyBuderus app logic. First you can change the schedule of one selected day and then saving it. Than copying to all other days.

> * Please try to change one time per block, and log states before and after write, s we can see if a `reject` ignores the values or writes them.

I have done it - the recected days are not written

**Comment by tp1de** [comment]
I tested -test.10 version with changing worddays (MO-FR) - 5 days  are 3 telegrams.
Out of 10 times testing only one was without errors.

**Comment by MichaelDvP** [comment]
Oh, fixed the delay now.
> And don't forget that read request (raw) give errors sometimes as well:
2024-11-23 12:17:03.575 I 56: [command] Called command system/send (Sende EMS-Telegramm) with value 0B 90 FF 3C 0C 01 C4
2024-11-23 12:17:25.174 E 57: [telegram] Last Tx Read operation failed after 3 retries. Ignoring request: 0B 10 FF 00 01 C4 0C

The error is not from that telegram sent 22 sec before. Also does not look like a read. Please send complete logs to see the context. Snipplets are mostly not usefull.

**Comment by tp1de** [comment]
With 1 second delay the number of tx write errors are a bit lower but there is no general change.

**Comment by Tim19791** [comment]
Hello,
Are there instructions on how I can get the entities, switchProgram A and B: hc1 from the RC310 into the iobroker?
Software is v3.7.2. Retrieved from https://github.com/MichaelDvP/EMS-ESP32/releases
 Unfortunately these entities do not exist, what else do I need to do?
Thanks!

Hallo,
gibt es eine Anleitung wie ich die  Entitäten,   switchProgram A und B: hc1 von der RC310 in den iobroker bekomme?
Software ist v3.7.2. Von https://github.com/MichaelDvP/EMS-ESP32/releases
 Leider sind diese Entitäten nicht vorhanden, was muss ich noch tun?
Danke!

**Comment by tp1de** [comment]
You need to install the latest test version from Michael .
The active switchProgram is then within switchprog.
The active version depends on program (1 or 2) and type ( Level or Absolute).

**Comment by tp1de** [comment]
@proddy @MichaelDvP   I recommend to release the switchPrograms to dev version.

The tx write errors which happen sometimes are not related to switchPrograms.
(I updated my ioBroker adapter and my node-red flow to issue the updates 3 times with some timedelay. This works)

**Comment by proddy** [comment]
ok. we would need some documentation too on how to use it and it's limitations. not everyone will use HA and NodeRed

**Comment by MichaelDvP** [comment]
It's mostly self explaining and in line with other entities, except that value is a json and build at publish time (stored only the raw telegram of 84 bytes).
Api shows entity `thermostat/hc1/switchprog` with type json and value as json array, attribute filtering works.
```
{"name":"switchprog",
"fullname":"hc2 switchprogram",
"circuit":"hc2",
"type":"json",
"value":[{"no":0,"day":"mo","time":"05:50","mode":"on"},{"no":1,"day":"mo","time":"21:30","mode":"off"},{"no":2,"day":"tu","time":"05:50","mode":"on"},{"no":3,"day":"tu","time":"21:30","mode":"off"},{"no":4,"day":"we","time":"05:50","mode":"on"},{"no":5,"day":"we","time":"21:30","mode":"off"},{"no":6,"day":"th","time":"05:50","mode":"on"},{"no":7,"day":"th","time":"21:30","mode":"off"},{"no":8,"day":"fr","time":"05:50","mode":"on"},{"no":9,"day":"fr","time":"21:30","mode":"off"},{"no":10,"day":"sa","time":"06:20","mode":"on"},{"no":11,"day":"sa","time":"21:30","mode":"off"},{"no":12,"day":"su","time":"07:00","mode":"on"},{"no":13,"day":"su","time":"21:30","mode":"off"}],
"readable":true,
"writeable":true,
"visible":true}
```
In mqtt this is published as extra topic `thermostat_switchprog` to not blow up the thermostat_data.
Command is as expected `cmd: switchprog` and `value:` as jsonObject `{  } `for a single time or as jsonArray `[{ },{ },..  ]`for many switchtimes at once. Writing all at once (4 telegram parts for 84 bytes) seems to be randomly rejected by thomas' RC310, on my RC35 it always work.

Web shows this entity like a command without value, only an arrow. 
![grafik](https://github.com/user-attachments/assets/300416a3-3a69-4842-8773-4fa306cb4aef)
You can enter a json in the input field. 
![grafik](https://github.com/user-attachments/assets/dc221b88-9bce-42e8-a07c-3b0193f4a747)
If we want to edit the switchprog in web, this should go to an extra page. It has too much inputs fields.

Should work for RC35, RC20, RC3x0 and some junkers, but tested only for RC35 and RC310.

RC3x0 holiday modes are not implemented yet, but i want to do it similar as json and publish as own topic thermostat_holidays.

**Comment by Marcwa19197** [comment]
Would be great to see this feature added!
Also one note, you can choose between "levels" (2-temperature level, one for day and one for night) and also you can choose a switchprogram with exact temperatures.

Using exact temperatures you need to also define the "Absenkschwelle" under which temperature nightmode / reduced mode will be activated. This value is also missing right now in EMS-ESP, but if you plan to include the timing settings you will also need to be able to set this value (if you use "exact temperatures", not 2-level temperatures).

![IMG_7194](https://github.com/user-attachments/assets/8e976bae-d4eb-4c8e-87cb-355fd3e9a5f2)
(its in german, sorry :-))

**Comment by tp1de** [comment]
The "Absenkschwelle" is defined in hcx reducetemp.
Btw this is active on Levels as well as Absolute

**Comment by Marcwa19197** [comment]
On my side, using a RC310 im not able to set the "Absenkschwelle" using 2-Level Mode, only while using absolute mode. 

Also i do not have the "reducetemp" entity... The value set on "Absenkschwelle" tells the switchprogram under whoch value reducetemp / reduce mode shoule be active. 

For example if i set Absenkschwelle to 17°C i can define a switchprogram starting at 8pm to 16°C then starting at 8pm heating will be go into reducemode (or will be off, depending on the setting)

**Comment by tp1de** [comment]
You are not using the hc controlmode depending on outside temp, isn't it?
Do you have an outside temp sensor installed?

**Comment by Marcwa19197** [comment]
I do use with outdoortempsensor and with room-influence.
See attached.

German:
Ich nutze die außengeführte Regelung mit Raumeinfluss (RC100 in der Wohnung), ich bin der Meinung, dass man die Absenkschwelle nur setzen kann, wenn man mit der "frei einstellbaren Temperatur" im Zeitprogramm arbeitet. Dann wird unter der eingestellten Temperatur der Absenkschwelle der absenkbetrieb aktiv (der wiederum je nach einstellung entweder einen reduzierten Betrieb unter einer eingestellten Außentemperatur fährt oder die Heizung komplett abschaltet). So steht es auch in der Anleitung von Buderus. War auch für mich anfangs verwirrend, ich konnte es aber gegentesten.

Stelle ich im Zeitprogramm eine frei wählbare Temperatur unter der Absenkschwelle ein, geht die Heizung ab dieser Zeit in den "Nacht/Absenkbetrieb". Oberhalb der Schwelle läuft sie ganz normal weiter.
Ich habe bei mir die Absenkschwelle auf 19 °C und ab 23:30 Uhr auf 17 °C im Zeitprogramm gestellt, "Absenkbetrieb unter" ist auf 0 °C Außentemperatur gestellt. Das hat zur Folge, dass ab 23:30 die Heizung, sofern es über 0 °C Außentemperatur ist, komplett aus bleibt. Ab unter 0 °C (und nach 23:30) wird im reduzierten Betrieb mit Solltemperatur 17 °C geheizt

[Thermostat_RC310_RC3*0, Moduline 3000_1010H, CW400, Sense II, HPC410.csv](https://github.com/user-attachments/files/18091965/Thermostat_RC310_RC3.0.Moduline.3000_1010H.CW400.Sense.II.HPC410.csv)

**Comment by Marcwa19197** [comment]
> "reducetemp" and " noreducetemp" define outside temps for hc controls. You are looking for "Absenkschwelle" for room temperature, which isn't defined yet within ems-esp.

Yes, this is what i wanted to know. But it makes sense, if the switchprogram functionality will be added to also add the "Absenkschwelle" because it is needed for the absolute switchprogram.

**Comment by tp1de** [comment]
So richtig verstehe ich deine Regelungsparameter nicht. Ich habe einen RC310 mit Außentemperatur-geführter Regelung mit Raumeinfluss und benutze Levels.

Dann kann ich mit reducetemp die Außentemperaturschwelle setzen unter der der Absenkbetrieb aktiviert ist und mit noreducetemp die Schwelle wo durchgeheizt wird. 

Ich bin aber aktuell nicht zuhause um zu testen.

**Comment by Marcwa19197** [comment]
Genau, an der Dämmung sind wir dran :-).. anderes Thema...

Wir verwechseln hier die Parameter, es gibt nämlich insgesamt drei:
- "hcx no reduce below temperature" --> Durchheizen auf Tagestemperatur
- "hcx off/reduce switch temperature" --> Ab unter X °C reduzierter Betrieb
- "Absenkschwelle" --> Der Parameter fehlt aktuell im EMS-ESP. Er wird benötigt um im Zeitprogramm bei "frei einstellbarere Temperatur" zu sagen ab welcher Uhrzeit der reduzierte Betrieb erfolgen kann. Da man im frei einstellbaren Zeitprogramm nur eine Temperatur oder "Aus" im Programm einstellen kann. Stelle ich nun den fehlenden Parameter "Absenkschwelle" auf z.b. 19 °C und im Zeitprogramm dann ab 23:30 Uhr auf "unter 19 °C" weiss das RC310 dass es ab dieser Zeit in den "reduzierten / Nachtbetrieb" laufen soll... Ist etwas schwierig zu erklären, vor allem noch auf englisch. Aber schaue es dir gerne mal bei Gelegenheit an. 

Die Absenkschwelle, die ich meine, wird nämlich im Bedienermenü Heizung -> Absenkschwelle eingestellt und ist nur sichtbar, wenn auf "frei einstellbare Temperatur" gestellt wurde. Bei dem 2-Temperatur-Niveau wird diese nicht benötigt, da ich da im Zeitprogramm einfach nur sage wann "Tagbetrieb" und wann "Nacht/reduzierter" Betrieb.

In der Print-Variante des Handbuchs steht folgendes (in der Online-Anleitung wurde der Absatz leider anders beschrieben):
"Temperatureinstellungen über das Zeitprogramm anpassen.
Wenn für einen Zeitabschnitt die Temperatur unterhalb der eingestellten Absenkschwelle liegt, ist in diesem Zeitabschnitt der Absenkbetrieb aktiv (z. B. eingeschränkter Heizbetrieb oder Heizung aus)."

Sorry for writing in german... complicated topic :D

**Comment by tp1de** [comment]
I added a new GH issue for the missing entity: #2288
